### ISSUE TITLE
Call the env hook only on PWD change in fish?

### ISSUE THREAD

**Issue by Bannerets** [ocaml/opam#5193]
Executing `opam env` on every prompt adds a bit of annoying delay (70-100 ms). In fish, it is possible to execute the command only when the current directory changes:

```fish
set -gx OPAMNOENVNOTICE true;
function __opam_env_export_eval --on-variable PWD
    eval (opam env --shell=fish --readonly 2> /dev/null);
end
```

instead of the current

```fish
set -gx OPAMNOENVNOTICE true;
function __opam_env_export_eval --on-event fish_prompt;
    eval (opam env --shell=fish --readonly 2> /dev/null);
end
```

Would that work fine for opam, or does it depend on more things than just the current directory? Maybe if the local switch was removed, I'm not sure.

It looks like zsh has the same functionality with `chpwd` / `chpwd_functions` instead of `precmd`. There's `cwdcmd` in tcsh.

**Comment by emillon** [comment]
Yes I noticed the same issue with zsh. A side effect is that when you're in a directory with a local switch, there's no way to temporarily override the switch to a global one. If the hook was attached to `cd`, it would be possible to first run `opam switch some-global-switch`.

**Comment by dra27** [comment]
> A side effect is that when you're in a directory with a local switch, there's no way to temporarily override the switch to a global one.

`opam exec --switch="foo" -- <cmd>` should work?

**Comment by emillon** [comment]
It's not always working. I suppose it's a different issue, but for example if `active` is the current active switch, the following does not print what I would expect:

    opam exec --switch other -- opam switch show

(it prints `active`)

**Comment by rjbou** [comment]
> Would that work fine for opam, or does it depend on more things than just the current directory? Maybe if the local switch was removed, I'm not sure.

@Bannerets We need to reload when user changes (global) switch too.
Can you run `opam env --shell=fish --debug-level=4` to see timestamps?

> `opam exec --switch other -- opam switch show`

@emillon in your case, it here a reason to not use directly `opam <opam-cmd> --swtch other` ?

**Comment by emillon** [comment]
I don't always have control about the nested opam command, for example it can be started by a script or another program. I have a workaround, in that case I can temporarily rename `_opam` to `_opam.bak` and it'll work. But I agree that `precmd` might be a bit too strong of a hook.

**Comment by Bannerets** [comment]
> Can you run opam env --shell=fish --debug-level=4 to see timestamps?

```
$ opam env --shell=fish --debug-level=4
00:00.068  FILE(config)           Read ~/.opam/config in 0.000s
00:00.069  CLI                    Parsing CLI version 2.1
00:00.069  GSTATE                 LOAD-GLOBAL-STATE @ /Users/user/.opam
00:00.069  SYSTEM                 LOCK /Users/user/.opam/lock (none => read)
00:00.069  FILE(config)           Read ~/.opam/config in 0.000s
00:00.070  CONFIG                 config-env
00:00.071  FILE(environment)      Read ~/.opam/4.14/.opam-switch/environment in 0.000s
00:00.071  FILE(environment)      Read ~/.opam/4.14/.opam-switch/environment in 0.000s
set -gx OPAM_SWITCH_PREFIX '/Users/user/.opam/4.14';
set -gx CAML_LD_LIBRARY_PATH '/Users/user/.opam/4.14/lib/stublibs:/Users/user/.opam/4.14/lib/ocaml/stublibs:/Users/user/.opam/4.14/lib/ocaml';
set -gx OCAML_TOPLEVEL_PATH '/Users/user/.opam/4.14/lib/toplevel';
set -gx PKG_CONFIG_PATH '/Users/user/.opam/4.14/lib/pkgconfig:';
set -gx PATH '/Users/user/.opam/4.14/bin' '/Users/user/.local/bin' '/Users/user/.cargo/bin' '/usr/local/bin' '/usr/bin' '/bin' '/usr/sbin' '/sbin' '/opt/X11/bin' '/Library/Frameworks/Mono.framework/Versions/Current/Commands' '/Applications/Wireshark.app/Contents/MacOS';
00:00.072  SYSTEM                 LOCK /Users/user/.opam/config.lock (none => none)
```

It looks like the "initialization" of opam itself takes 70 ms, e.g. `time opam > /dev/null` often takes 70 ms as well.
